freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/migration/AwsVariantMigrationActions.java

```diff
@@ -0,0 +1,119 @@
+
+    @Inject
+    private StackUpdater stackUpdater;
+
+    @Inject
+    private FreeIpaMetricService metricService;
+
+    @Bean(name = "CREATE_RESOURCES_STATE")
+    public Action<?, ?> createResources() {
+        return new AbstractAwsVariantMigrationAction<>(AwsVariantMigrationTriggerEvent.class) {
+
+            @Override
+            protected void doExecute(AwsVariantMigrationFlowContext context, AwsVariantMigrationTriggerEvent payload, Map<Object, Object> variables)
+                    throws Exception {
+                CreateResourcesRequest request = new CreateResourcesRequest(context.getCloudContext(), context.getCloudCredential(), context.getCloudStack(),
+                        payload.getHostGroupName());
+                sendEvent(context, request.selector(), request);
+            }
+
+        };
+    }
+
+    @Bean(name = "DELETE_CLOUD_FORMATION_STATE")
+    public Action<?, ?> deleteCloudFormation() {
+        return new AbstractAwsVariantMigrationAction<>(CreateResourcesResult.class) {
+            @Override
+            protected void doExecute(AwsVariantMigrationFlowContext context, CreateResourcesResult payload, Map<Object, Object> variables)
+                    throws Exception {
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(AwsVariantMigrationFlowContext context) {
+                return new DeleteCloudFormationRequest(context.getCloudContext(), context.getCloudCredential(), context.getCloudStack());
+            }
+        };
+    }
+
+    @Bean(name = "CHANGE_VARIANT_STATE")
+    public Action<?, ?> changeVariant() {
+        return new AbstractAwsVariantMigrationAction<>(DeleteCloudFormationResult.class) {
+            @Override
+            protected void doExecute(AwsVariantMigrationFlowContext context, DeleteCloudFormationResult payload, Map<Object, Object> variables)
+                    throws Exception {
+                if (payload.isCloudFormationTemplateDeleted()) {
+                    LOGGER.debug("Variant will be changed");
+                    stackUpdater.updateVariant(payload.getResourceId(), CloudConstants.AWS_NATIVE);
+                    LOGGER.info("Variant changed to AWS_NATIVE");
+                    metricService.incrementMetricCounter(MetricType.AWS_VARIANT_MIGRATION_SUCCESSFUL, context.getStack());
+                } else {
+                    LOGGER.info("Variant won't be changed because the CF template exists");
+                }
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(AwsVariantMigrationFlowContext context) {
+                return new StackEvent(AWS_VARIANT_MIGRATION_FINALIZED_EVENT.event(), context.getStack().getId());
+            }
+        };
+    }
+
+    @Bean(name = "AWS_VARIANT_MIGRATION_FAILED_STATE")
+    public Action<?, ?> migrationFailed() {
```

Comment:
it is a separated flow, how cloud I handle the failed step in FreeIPA?

Reply:
add operationid to the event, and if it's present in the event, you can fail it, please check for exmaple this: https://github.com/hortonworks/cloudbreak/blob/master/freeipa/src/main/java/com/sequenceiq/freeipa/flow/freeipa/downscale/action/FreeIpaDownscaleActions.java#L455

Reply:
Understood, Line added

Reply:
@lacikaaa why we should use this OperationService for storing flow information?
I haven't found the same approach in other services, if this is necessary a doc would be nice to have, but I think the flow engine should be enhanced in a common way instead of forcing a flow related operation logic/service.

Reply:
this was originally for tracking usersync (which is not flow based), but it seemed to be useful for other FreeIPA related operations, as the customers could track the state/progress of an operation like repair etc, as there is no event history for freeipa. so we are using it for most flow related operation too in fms.

Reply:
I have updated the failed branch to handle the operation id